### PR TITLE
Cross binutils

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

There's a lot more consistency now:
 - binutils binaries are always prefixed, when compiling from Darwin and Linux alike
 - binutils + cctools are always mixed together the same way if Darwin is the target

And there are tests to insure the darwin-targeting binutils+cctools mashups do build.

The main question is whether we want to so prefix the binutils derivation name, or adopt another convention distinguished from the existing `build != host` suffixes.

Builds on https://github.com/NixOS/nixpkgs/pull/25227, factored out of https://github.com/NixOS/nixpkgs/pull/25047

###### Things done

- http://hydra.nixos.org/eval/1359577

### CODE REVIEW

**Comment by Ericson2314** [comment]
@shlevy its hanging and won't let me request a review from you, so I'm doing it here.

**Comment by Ericson2314** [comment]
Two issues are:

 - My previously-merged triple parsing see the mingw stuff as ambiguous. A change in this PR forces that thunk causing `mingw` builds to fail. The triple can be fixed CC @taktoa

 - the `ios-cross` stuff is broken because `cc-wrapper` doesn't support the prefixing. The original PR fully fixes `cc-wrapper` for all cross, but some linux tests break because of missing attributes. I do have a commit that props of `cc-wrapper` a bit---enough for `ios-cross` to build---though adding it causes more churn.

File: pkgs/stdenv/linux/make-bootstrap-tools-cross.nix [hunk 1]
```diff
@@ -14,7 +14,7 @@ let
       float = "soft";
       withTLS = true;
       libc = "glibc";
-      platform = lib.systems.platforms.sheevaplug;
+      platform = pkgsNoParams.platforms.sheevaplug;
```
Comment (bjornfor) [inline pkgs/stdenv/linux/make-bootstrap-tools-cross.nix]: Didn't you just merge a PR to master that did the opposite change (pkgsNoParame -> lib.systems)?

**Comment by taktoa** [comment]
@Ericson2314 I never bothered to fix the parser for mingw32, and am fairly busy with other stuff at the moment. It should be easy though.

File: pkgs/stdenv/linux/make-bootstrap-tools-cross.nix (hunk 1 above)
Comment (Ericson2314) [inline pkgs/stdenv/linux/make-bootstrap-tools-cross.nix]: Ah all my rebasing runs the risk of stuff like this. Good catch.

**Comment by Ericson2314** [comment]
Fixed up mingw32 so there's no longer any regressions.

**Comment by Ericson2314** [comment]
Ok rebased this on merged (and remerged, welp) other PRs. Also re-ran the hydra job to ensure everything is ship-shape (so far, so good).

**Comment by Ericson2314** [comment]
Ok I think the remaining builds are stalled or something, and I didn't operate hydra correctly to restart. This all worked before I rebased, and "by parametricity" the remaining builds should also work fine, and nobody reviewed up or down while I was out of town, so... merging!

**Pull request merged**